GVFS/GVFS.Common/Cleanup/GitCleanupQueue.cs

```diff
@@ -0,0 +1,103 @@
+﻿using GVFS.Common.Tracing;
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Threading;
+
+namespace GVFS.Common.Cleanup
+{
+    public class GitCleanupQueue
+    {
+        private GVFSContext context;
+        private BlockingCollection<GitCleanupStep> queue = new BlockingCollection<GitCleanupStep>();
+        private CancellationTokenSource cancellationToken = new CancellationTokenSource();
+        private GitCleanupStep currentStep;
+
+        public GitCleanupQueue(GVFSContext context)
+        {
+            this.context = context;
+            Thread cleanupWorker = new Thread(() => this.RunQueue());
+            cleanupWorker.Name = "CleanupWorker";
+            cleanupWorker.Start();
+        }
+
+        public void Enqueue(GitCleanupStep step)
+        {
+            this.queue.Add(step);
+        }
+
+        public void Stop()
+        {
+            this.cancellationToken.Cancel();
+            this.currentStep?.Stop();

```

Comment:
Looking at the api/source we can probably replace the cancellationtoken in the class with a call to CompleteAdding().  That would make the class even smaller / sleeker!  We'd have to check what happens if you try to add in that scenario (does it throw an exception) and make sure we handle that case correctly.

Reply:
If we drop the cancellation token, then we just need to replace it with a `Stopping` boolean again. I'll add the `CompleteAdding()` to avoid adding some more, but I think we should keep the cancellation token.

Reply:
> I'll add the CompleteAdding() to avoid adding some more, but I think we should keep the cancellation token.

What scenario do we still need the cancellation token for? We can check `this.queue.IsAddingComplete` to see if we've called `CompleteAdding()` (and stop running steps if `IsAddingComplete` is true).

Reply:
@derrickstolee pointed this out to me.

There could already be Steps already waiting in the queue.  By using the cancellation token we won't try to read them.

Reply:
> There could already be Steps already waiting in the queue. By using the cancellation token we won't try to read them.

I think if we we switch to an if rather than a try\catch we should be able to avoid that scenario:

Something like this:

```
while (true)
{
    if (!this.queue.TryTake(out this.currentStep, Timeout.Infinite) || this.queue.IsAddingComplete)
    {
        // Stop has been requested
       return;
```

Reply:
@wilbaker CompleteAdding() prevents new items from being added.  I would expect it to break out of the TryTake so that would wait there indefinitely right?

I like the approach of cancelling the TryTake which causes the background thread to end gracefully.

Reply:
That'll do! thanks.

Reply:
Good point, @jeschu1. Need to think about that. I know that the unit tests would wait until that `TryTake` completes if we didn't call `Stop()` directly. They seem to stop fine now, but I'll do some digging into the `TryTake()`.

Reply:
The `CompleteAdding` will [cause the `TryTake` to return `false`](https://stackoverflow.com/a/21195735).

Reply:
> @wilbaker CompleteAdding() prevents new items from being added. I wouldn't expect it to break out of the TryTake so that would wait there indefinitely right?

`TryTake` will break out and return false if `CompleteAdding()` is called while it's waiting.

Internally the BlockingCollection uses a CancellationToken to break out of TryTakes that are waiting (without a user provided cancellation token).

*From TryTake*
https://referencesource.microsoft.com/#System/sys/system/collections/concurrent/BlockingCollection.cs,712

```
// set the combined token source to the combinedToken paramater if it is not null (came from GetConsumingEnumerable)
CancellationTokenSource linkedTokenSource = combinedTokenSource;
try
{
    waitForSemaphoreWasSuccessful = m_occupiedNodes.Wait(0);
    if (waitForSemaphoreWasSuccessful == false && millisecondsTimeout != 0)
    {
        // create the linked token if it is not created yet
        if (combinedTokenSource == null)
            linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken,
                                                                                m_ConsumersCancellationTokenSource.Token);
        waitForSemaphoreWasSuccessful = m_occupiedNodes.Wait(millisecondsTimeout, linkedTokenSource.Token);
    }
}
//The collection became completed while waiting on the semaphore.
catch (OperationCanceledException)
{
    if (cancellationToken.IsCancellationRequested)
        throw new OperationCanceledException(SR.GetString(SR.Common_OperationCanceled), cancellationToken);

    return false;
}
```

*From CompleteAdding:*
https://referencesource.microsoft.com/#System/sys/system/collections/concurrent/BlockingCollection.cs,1516
```
if (Interlocked.CompareExchange(ref m_currentAdders, observedAdders | COMPLETE_ADDING_ON_MASK, observedAdders) == observedAdders)
{
    spinner.Reset();
    while (m_currentAdders != COMPLETE_ADDING_ON_MASK) spinner.SpinOnce();

    if (Count == 0)
    {
        CancelWaitingConsumers();
    }
```

We have used this approach elsewhere VFS4G and haven't seen shutdown issues (e.g. `GitIndexGenerator.WriteAllEntries` and `FileSystemVirtualizer.ExecuteFileOrNetworkRequest`)

Reply:
Thanks @wilbaker for the education there, yes that will definitely work nicely and eliminate the variable, nice! :-)